File: src/library/ctest/src/prho.f
c     Was double precision function prho(n, is, ifault).
c     Changed to subroutine by KH to allow for .Fortran interfacing.
c     Also, change `prho' in the code to `pv'.
c     And, fix a bug.
c
      subroutine prho(n, is, pv, ifault)
c
c	 Algorithm AS 89   Appl. Statist. (1975) Vol.24, No. 3, P377.
c
c	 To evaluate the probability of obtaining a value greater than or
c	 equal to is, where is=(n**3-n)*(1-r)/6, r=Spearman's rho and n
c	 must be greater than 1
c
c     Auxiliary function required: ALNORM = algorithm AS66
C     MM: Rather use  1 - pnorm()  {only AFTER translation to C !}
c
c     KH/MM add/beg
      implicit none
      integer n, is, ifault
      double precision pv
c     KH add/end
      integer l(6)
      integer i,m,mt,js,n1,nn,nfac,ifr,ise
C External
      double precision alnorm
C     MM add/end
      double precision zero, one, two, b, x, y, u, six,
     $	c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12
      data zero, one, two, six /0.0d0, 1.0d0, 2.0d0, 6.0d0/
      data	c1,	c2,	c3,	c4,	c5,	c6,
     $		c7,	c8,	c9,    c10,    c11,    c12/
     $	0.2274d0, 0.2531d0, 0.1745d0, 0.0758d0, 0.1033d0, 0.3932d0,
     $	0.0879d0, 0.0151d0, 0.0072d0, 0.0831d0, 0.0131d0, 0.00046d0/
c
c	 Test admissibility of arguments and initialize
c
      pv = one
      ifault = 1
      if (n .le. 1) return
      ifault = 0
      if (is .le. 0) return
      pv = zero
      if (is .gt. n * (n * n -1) / 3) return
      js = is
      if (js .ne. 2 * (js / 2)) js = js + 1
      if (n .gt. 6) goto 6
c
c	 Exact evaluation of probability
c
      nfac = 1
      do 1 i = 1, n
	nfac = nfac * i
	l(i) = i
    1 continue
      pv = one / dble(nfac)
c     KH mod/beg
c     This was `.ne.' in the code but `.eq.' in the paper
      if (js .eq. n * (n * n - 1) / 3) return
c     KH mod/end
      ifr = 0
      do 5 m = 1,nfac
	ise = 0
	do 2 i = 1, n
	  ise = ise + (i - l(i)) ** 2
    2	continue
	if (js .le. ise) ifr = ifr + 1
	n1 = n
    3	mt = l(1)
	nn = n1 - 1
	do 4 i = 1, nn
	  l(i) = l(i + 1)
    4	continue
	l(n1) = mt
	if (l(n1) .ne. n1 .or. n1 .eq. 2) goto 5
	n1 = n1 - 1
	if (m .ne. nfac) goto 3
    5 continue
      pv = dble(ifr) / dble(nfac)
      return
c
c     Evaluation by Edgeworth series expansion
c
    6 b = one / dble(n)
      x = (six * (dble(js) - one) * b / (one / (b * b) -one) -
     $	one) * sqrt(one / b - one)
      y = x * x
      u = x * b * (c1 + b * (c2 + c3 * b) + y * (-c4
     $	+ b * (c5 + c6 * b) - y * b * (c7 + c8 * b
     $	- y * (c9 - c10 * b + y * b * (c11 - c12 * y)))))
c
c     Call to algorithm AS 66
c
      pv = u / exp(y / two) + alnorm(x, .true.)
      if (pv .lt. zero) pv = zero
      if (pv .gt. one) pv = one
      return
      end

File: src/library/ctest/src/prho.c
/*
 *  R : A Computer Language for Statistical Data Analysis
 *  Copyright (C) 2000		R Development Core Team
 *  based on AS 89 (C) 1975 Royal Statistical Society
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */
#include "R_ext/Mathlib.h"

/* Was
	double precision function prho(n, is, ifault)

 Changed to subroutine by KH to allow for .Fortran interfacing.
 Also, change `prho' argument in the code to `pv'.
 And, fix a bug.

 From R ver. 1.1.x [March, 2000] by MM:
 - Translate Fortran to C
 - use pnorm() instead of less precise alnorm().
 - new argument lower_tail --> potentially increased precision in extreme cases.
*/
void prho(int *n, int *is, double *pv, int *ifault, int *lower_tail)
{
/*	Algorithm AS 89	  Appl. Statist. (1975) Vol.24, No. 3, P377.

	To evaluate the probability  Pr[ S >= is ]
	{or Pr [ S < is]  if(lower_tail) }, where

	S   = (n^3 - n) * (1-R)/6,
	is  = (n^3 - n) * (1-r)/6,
	R,r = Spearman's rho (r.v. and observed),  and	n >= 2
*/

    /* Edgeworth coefficients : */
    const double
	c1 = .2274,
	c2 = .2531,
	c3 = .1745,
	c4 = .0758,
	c5 = .1033,
	c6 = .3932,
	c7 = .0879,
	c8 = .0151,
	c9 = .0072,
	c10= .0831,
	c11= .0131,
	c12= 4.6e-4;

    /* Local variables */
    double b, u, x, y;
    int l[6];
    int nfac, i, m, n1, js, mt, ifr, ise;

    /* Test admissibility of arguments and initialize */
    *pv = *lower_tail ? 0. : 1.;
    if (*n <= 1) { *ifault = 1; return; }

    *ifault = 0;
    if (*is <= 0) return;/* with p = 1 */

    *pv = 1 - *pv;
    n1 = *n * (*n * *n - 1) / 3;
    if (*is > n1)
	return;

    js = *is;
    if(js % 2)/* if js is not even, make it */
	++js;


    if (*n <= 6) { /* 2 <= n <= 6 : Exact evaluation of probability */

	nfac = 1;
	for (i = 1; i <= *n; ++i) {
	    nfac *= i;
	    l[i - 1] = i;
	}
	/* KH mod next line: was `!=' in the code but `.eq.' in the paper */
	if (js == n1) {
	    ifr = 1;
	}
	else {
	    ifr = 0;
	    for (m = 0; m < nfac; ++m) {
		ise = 0;
		for (i = 0; i < *n; ++i) {
		    n1 = i + 1 - l[i];
		    ise += n1 * n1;
		}
		if (js <= ise)
		    ++ifr;

		n1 = *n;
		do {
		    mt = l[0];
		    for (i = 1; i < n1; ++i)
			l[i - 1] = l[i];
		    --n1;
		    l[n1] = mt;
		} while (mt == n1+1 && n1 > 1);
	    }
	}
	*pv = (*lower_tail ? nfac-ifr : ifr) / (double) nfac;
    } /* exact for n <= 6 */

    else { /* n >= 7 :	Evaluation by Edgeworth series expansion */

	y = (double) (*n);
	b = 1 / y;
	x = (6. * (js - 1) * b / (y * y - 1) - 1) * sqrt(y - 1);
	y = x * x;
	u = x * b * (c1 + b * (c2 + c3 * b) +
		     y * (-c4 + b * (c5 + c6 * b) -
			  y * b * (c7 + c8 * b -
				   y * (c9 - c10 * b + y * b * (c11 - c12 * y))
			      )));
	y = u / exp(y / 2.);
	*pv = (*lower_tail ? -y : y) +
	    pnorm(x, 0., 1., *lower_tail, /*log_p = */LFALSE);
	/* above was call to alnorm() [algorithm AS 66] */
	if (*pv < 0) *pv = 0.;
	if (*pv > 1) *pv = 1.;
    }
    return;
} /* prho */

